javascript/ql/src/Expressions/MissingAwait.ql

```diff
@@ -0,0 +1,73 @@
+/**
+ * @name Missing await
+ * @description Using a promise without awaiting its result can lead to unexpected behavior.
+ * @kind problem
+ * @problem.severity warning
+ * @id js/missing-await
+ * @tags correctness
+ * @precision high
+ */
+
+import javascript
+
+/**
+ * Holds if `call` is a call to an `async` function.
+ */
+predicate isAsyncCall(DataFlow::CallNode call) {
+  // If a callee is known, and all known callees are async, assume all
+  // possible callees are async.
+  forex(Function callee | call.getACallee() = callee | callee.isAsync())
+}
+
+/**
+ * Holds if `node` is always a promise.
+ */
+predicate isPromise(DataFlow::SourceNode node, boolean nullable) {
+  isAsyncCall(node) and
+  nullable = false
+  or
+  not isAsyncCall(node) and
+  node.asExpr().getType() instanceof PromiseType and
+  nullable = true
+}
+
+/**
+ * Holds the result of `e` is used in a way that doesn't make sense for Promise objects.
+ */
+predicate isBadPromiseContext(Expr expr) {
+  exists(BinaryExpr binary |
+    expr = binary.getAnOperand() and
+    not binary instanceof LogicalExpr and
```

Comment:
I assume you have tried flagging left operands of logical expressions and concluded that people tend to actually do that (even though it's pointless)?

What about `expr ? ... : ...` and `if (expr) ...`? (And perhaps even `base[expr]`?)

Reply:
Also, `for(x in expr) ...` (though probably not `for(x of expr)`, since that actually produces a run-time error).

Reply:
Added a few more cases. The left operand of logical expressions are handled in a separate case below the `exists`.